internal/model/asset.go

```
@@ -0,0 +1,80 @@
+package model
+
+// AssetType defines the type of asset
+type AssetType string
+
+// Supported asset types
+const (
+	AssetTypeChart    AssetType = "chart"
+	AssetTypeInsight  AssetType = "insight"
+	AssetTypeAudience AssetType = "audience"
+)
+
+// BaseAsset contains common fields for all assets
+type BaseAsset struct {
+	ID          string    `json:"id"`
+	Type        AssetType `json:"type"`
+	Description string    `json:"description"`
+	CreatedAt   int64     `json:"created_at"`
+	UpdatedAt   int64     `json:"updated_at"`
+}
+
+// Chart represents a chart asset
+type Chart struct {
+	BaseAsset `json:",inline"`
+	Title     string        `json:"title"`
+	XAxisKey  string        `json:"x_axis_key"`
+	YAxisKey  string        `json:"y_axis_key"`
+	Data      []interface{} `json:"data"`
+}
+
+// Insight represents an insight asset
+type Insight struct {
+	BaseAsset `json:",inline"`
+	Title     string `json:"title"`
+	Content   string `json:"content"`
+}
+
+// Audience represents an audience segment
+type Audience struct {
+	BaseAsset `json:",inline"`
+	Gender    string   `json:"gender"` // "Male", "Female", or ""
+	Countries []string `json:"countries"`
+	AgeMin    int      `json:"age_min"`
+	AgeMax    int      `json:"age_max"`
+	MinHours  int      `json:"min_hours_daily_social_media"`
+	MinBuys   int      `json:"min_purchases_last_month"`
+}
+
+// Asset is a union type that holds any asset
+type Asset struct {
+	Chart    *Chart    `json:"chart,omitempty"`
```

Suppose the following valid asset definition:
(I also took a look at your similar test case in `asset_test.go` line 79).

```go
asset := &model.Asset{
    Chart: &model.Chart{
        BaseAsset: model.BaseAsset{
            ID: "chart-1",
            Type: model.AssetTypeChart,
        },
    },
    Audience: &model.Audience{
        BaseAsset: model.BaseAsset{
            ID: "audience-1",
            Type: model.AssetTypeAudience,
        },
    },
    Insight: &model.Insight{
        BaseAsset: model.BaseAsset{
            ID: "insight-1",
            Type: model.AssetTypeInsight,
        },
    },	
}
```

The POST `/api/users/{id}/favorites` controller receives a request with this in its request body. This asset will be stored as favorited by user {id}.

How many favorite assets does this user have? What are your thoughts on this approach? Does it achieve the desired behavior or address the underlying issue?

---

The user would have 1 favorite containing all 3 asset types at the same time, which is invalid. The `Asset` union type has no validation to enforce that exactly one type should be set. `GetBaseAsset()` would just return the first non-nil field (`Chart` in this case), silently ignoring the others. I'd use the interface-based approach which makes invalid states impossible by design